### PR TITLE
[WIP] Short-term UX improvements for HTTPS-Everywhere onion support

### PR DESCRIPTION
**Pull request by vporton**
### List related PRs if any

https://github.com/EFForg/https-everywhere/pull/18994

### List related issues if any

https://github.com/acatarineu/tor-browser/issues/10

### Notes

I've implemented "A panel that allows users to view all the currently enabled onion rules and update channels."

Weirdly it shows only 3 rules. Is it an error? I use code from https://github.com/EFForg/https-everywhere/pull/18994 to fetch the rules. Is that code erroneous?

UX of that table in the popup is bad (the data is long and does not fit into the dialog window size). Please advice how to improve it.

### CODE REVIEW

**Comment by pipboy96** [comment]
@vporton Please don't mix unrelated changes in a single PR. Instead rebase your UX changes to the current master branch and keep the Python version upgrade as a separate PR.

File: chromium/pages/popup/onion-rules.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  /**
+ * Fill in content into the popup on load
+ */
+document.addEventListener("DOMContentLoaded", function () {
+  // auto-translate all elements with i18n attributes
+  var elem = document.querySelectorAll("[i18n]");
+  for (var i=0; i < elem.length; i++) {
+    elem[i].innerHTML = chrome.i18n.getMessage(elem[i].getAttribute("i18n"));
+  }
+
```
Comment (pipboy96) [inline chromium/pages/popup/onion-rules.js:14]: `translation.js` already handles this.

```suggestion
```

File: chromium/pages/popup/onion-rules.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
+function e(id) {
+return document.getElementById(id);
+}
+  
+  /**
```
Comment (pipboy96) [inline chromium/pages/popup/onion-rules.js:5]: Please use correct indentation here.

```suggestion
/**
```

File: chromium/pages/popup/onion-rules.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  }
+
+  showRules();
+});
+
+function _createTD(text) {
+  var td = document.createElement('td');
+  var textNode = document.createTextNode(text);
+  td.appendChild(textNode);
+  return td;
```
Comment (pipboy96) [inline chromium/pages/popup/onion-rules.js:22]: There is a simpler way to do that:

```suggestion
  var td = document.createElement('td');
  td.textContent = text;
  return td;
```

**Comment by pipboy96** [comment]
Otherwise seems good except of Python-related changes.

**Comment by vporton** [comment]
> If you could walk back the `python3.8` changes in this PR and just focus the UX I can pull down and review this.

Done.

> For any sweeping UX changes I have to consult with our EFF design folks. Once the the python changes have been reverted here I can pull this PR down and take a look.

Please do.

Please also look at https://github.com/acatarineu/tor-browser/issues/10#issuecomment-649340903.

File: chromium/pages/popup/onion-rules.html [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  <body>
+    <h1>Enabled Onion Rules</h1>
+    <table>
+        <thead>
+            <tr><th>From regex</th><th>Scope regex</th><th>Host</th><th>To</th></tr>
+        </thead>
+        <tbody id="onionTable"></tbody>
+    </table>
+  </body>
+  </html>
```
Comment (pipboy96) [inline chromium/pages/popup/onion-rules.html:23]: Please add a newline at the end of this file.

File: chromium/pages/popup/onion-rules.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+      var rule = rules[i];
+      var tr = document.createElement("tr");
+      tr.appendChild(_createTD(rule.from_regex));
+      tr.appendChild(_createTD(rule.scope_regex));
+      tr.appendChild(_createTD(rule.host));
+      tr.appendChild(_createTD(rule.to));
+      tbody.appendChild(tr);
+    }
+  });
+}
```
Comment (pipboy96) [inline chromium/pages/popup/onion-rules.js:38]: Please add a newline at the end of this file.

File: chromium/pages/popup/onion-rules.html (hunk 4 above)
Comment (vporton) [reply chromium/pages/popup/onion-rules.html]: Done.

File: chromium/pages/popup/onion-rules.js (hunk 5 above)
Comment (vporton) [reply chromium/pages/popup/onion-rules.js]: Done.

File: chromium/pages/popup/onion-rules.html [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
```
Comment (pipboy96) [inline chromium/pages/popup/onion-rules.html:24]: Please remove these two spaces.

```suggestion

```

Comment (vporton) [reply chromium/pages/popup/onion-rules.html]: Done.

**Comment by vporton** [comment]
I added `chromium/background-scripts/urlbarprovidertoronion.js`, but when I try to load it, I get an error. Apparently it is related with extension signing, because other JS files in the same folder load fine. Please help!!!

`UrlbarProviderTorOnion = require('./urlbarprovidertoronion').UrlbarProviderTorOnion`

`Error: module: ./urlbarprovidertoronion not found.`

Also, when I type, `require('./urlbarprovidertoronion').UrlbarProviderTorOnion` in console, it works fine. It is some mystery related with extension signing. Explain me how to get around of this in Tor.

**Comment by pipboy96** [comment]
@vporton What are the contents of `urlbarprovidertoronion.js`?

**Comment by vporton** [comment]
> @vporton What are the contents of `urlbarprovidertoronion.js`?

@pipboy96 
```
'use strict';

(function (exports) {
    class UrlbarProviderTorOnion {
        get name() {
            return "UrlbarProviderTorOnion";
        }
        get type() {
            UrlbarUtils.PROVIDER_TYPE.SUGGESTION;
        }
        isActive(queryContext) {
            return true;
        }
        isRestricting(queryContext) {
            return false;
        }
        /**
            * Starts querying.
            * @param {UrlbarQueryContext} queryContext The query context object
            * @param {function} addCallback Callback invoked by the provider to add a new
            *        result. A UrlbarResult should be passed to it.
            * @note Extended classes should return a Promise resolved when the provider
            *       is done searching AND returning results.
            * @abstract
            */
        startQuery(queryContext, addCallback) {
            addCallback(
            this,
            new UrlbarResult(
                UrlbarUtils.RESULT_TYPE.URL,
                UrlbarUtils.RESULT_SOURCE.OTHER_LOCAL, // or OTHER_NETWORK?
                { icon: null,
                url: "zzzzzzzzzzzzz",
                title: "",
                tags: [] }
            )
            );
        }
        cancelQuery(queryContext) { }
    }

    Object.assign(exports, {
        UrlbarProviderTorOnion
      });

})(typeof exports !== 'undefined' ? exports : require.scopes.urlbarprovidertoronion = {});
      
```

**Comment by pipboy96** [comment]
@vporton Try adding `urlbarprovidertoronion.js` to [manifest.json](https://github.com/EFForg/https-everywhere/blob/master/chromium/manifest.json):

```diff
     "background": {
         "scripts": [
             "background-scripts/bootstrap.js",
             "background-scripts/util.js",
             "wasm/https_everywhere_lib_wasm.js",
             "background-scripts/wasm.js",
             "background-scripts/update_channels.js",
             "background-scripts/update.js",
             "background-scripts/rules.js",
             "background-scripts/store.js",
             "external/pako-1.0.5/pako_inflate.min.js",
             "background-scripts/incognito.js",
             "background-scripts/ip_utils.js",
+            "background-scripts/urlbarprovidertoronion.js"
             "background-scripts/background.js"
         ]
     },
```

**Comment by vporton** [comment]
> @vporton Try adding `urlbarprovidertoronion.js` to [manifest.json](https://github.com/EFForg/https-everywhere/blob/master/chromium/manifest.json):

@pipboy96 I already did. But it does not work anyway.

As far as I remember, before adding to `manifest.json` it did not work at all, even from the console. Now it works from the console (see above) but not from `background-scripts/background.js`.

**Comment by pipboy96** [comment]
@vporton Please:
1. Run `console.log(require.scopes)` in console.
2. Add `console.log(require.scopes)` before `UrlbarProviderTorOnion = require('./urlbarprovidertoronion').UrlbarProviderTorOnion`.
3. Copy and paste the results here.

**Comment by vporton** [comment]
@pipboy96 

> @vporton Please:
> 
> 1. Run `console.log(require.scopes)` in console.

```
{…}
​
background: Object {  }
​
incognito: Object { onIncognitoDestruction: onIncognitoDestruction()
, state: {…} }
​
ip_utils: Object { parseIp: parseIp(), isIpInRange: isIpInRange(), isLocalIp: isLocalIp()
 }
​
rules: Object { settings: {…}, trivial_rule: {…}, Rule: Rule()
, … }
​
store: Object { initialize: initialize(), performMigrations: performMigrations()
 }
​
update: Object { applyStoredRulesets: applyStoredRulesets(), initialize: initialize(), getRulesetTimestamps: getRulesetTimestamps()
, … }
​
update_channels: Object { update_channels: (1) […] }
​
urlbarprovidertoronion: Object { UrlbarProviderTorOnion: UrlbarProviderTorOnion()
 }
​
util: Object { VERB: 1, DBUG: 2, INFO: 3, … }
​
wasm: Object { initialize: initialize(), RuleSets: RuleSets(), is_enabled: is_enabled()
 }
​
<prototype>: Object { … }
```

> 2. Add `console.log(require.scopes)` before `UrlbarProviderTorOnion = require('./urlbarprovidertoronion').UrlbarProviderTorOnion`.

```
{…}
​
background: Object {  }
​
incognito: Object { onIncognitoDestruction: onIncognitoDestruction()
, state: {…} }
​
ip_utils: Object { parseIp: parseIp(), isIpInRange: isIpInRange(), isLocalIp: isLocalIp()
 }
​
rules: Object { settings: {…}, trivial_rule: {…}, Rule: Rule()
, … }
​
store: Object { initialize: initialize(), performMigrations: performMigrations()
 }
​
update: Object { applyStoredRulesets: applyStoredRulesets(), initialize: initialize(), getRulesetTimestamps: getRulesetTimestamps()
, … }
​
update_channels: Object { update_channels: (1) […] }
​
urlbarprovidertoronion: Object { UrlbarProviderTorOnion: UrlbarProviderTorOnion()
 }
​
util: Object { VERB: 1, DBUG: 2, INFO: 3, … }
​
wasm: Object { initialize: initialize(), RuleSets: RuleSets(), is_enabled: is_enabled()
 }
​
<prototype>: Object { … }
```

> 3. Copy and paste the results here.

**Comment by pipboy96** [comment]
@vporton Thanks! Now, please:
1. Run `console.log(require.toString())` in console.
2. Add `console.log(require.toString())` before `UrlbarProviderTorOnion = require('./urlbarprovidertoronion').UrlbarProviderTorOnion`.
3. Copy and paste the results here.

**Comment by vporton** [comment]
> @vporton Thanks! Now, please:
> 
> 1. Run `console.log(require.toString())` in console.

```
function require(module) {
  if (module.startsWith('./') && require.scopes.hasOwnProperty(module.slice(2))) {
    return require.scopes[module.slice(2)];
  }
  throw new Error('module: ' + module + ' not found.');
}
```

> 2. Add `console.log(require.toString())` before `UrlbarProviderTorOnion = require('./urlbarprovidertoronion').UrlbarProviderTorOnion`.

```
function require(module) {
  if (module.startsWith('./') && require.scopes.hasOwnProperty(module.slice(2))) {
    return require.scopes[module.slice(2)];
  }
  throw new Error('module: ' + module + ' not found.');
}
```

> 3. Copy and paste the results here.

**Comment by pipboy96** [comment]
@vporton There are no visible issues here. @zoracon Can you take a look at this?

**Comment by cschanaj** [comment]
> A panel that allows users to view all the currently enabled onion rules and update channels. This would allow them to see all included securedrop rules right now, and in the future any other rule or update channel we decide to add.

Was there any discussion mentioning why such feature should be added to the extension instead of the Tor Browser itself? I am more inclined to think that each onion service should create a ATLAS for their onion service like (https://atlas.eff.org/index.html)  and possibly include these ATLAS as a bookmark in the TBB. In this way TBB would be able to _autocomplete_ the bookmarks from the omnibar natively.  

Firstly, the Tor Browser hide HTTPS Everywhere from its default UX arrangement. Less tech-savvy users have limited access to the features of HTTPS Everywhere inherently, including the proposed dashboard. This might introduce yet another UX issue for promoting the onion services. 

For HTTPS Everywhere users who use the extension without tor, it could be confusing to have a panel listing onion site they cannot reach without a dedicate setup.

**Comment by vporton** [comment]
I've reverted teh changes related to item 2 of the issue. Check if I correctly implemented the item 1.

**Comment by zoracon** [comment]
Closing due to the Onion experience being enhanced on Tor's side

**Pull request closed without merging**